### ISSUE TITLE
Uncaught TypeError on Json 'ACSURL' in Globalpay/Realex/Sagepay checkout modules

### ISSUE THREAD

**Issue by TerryMGray** [opencart/opencart#4711]
The 3d secure authentication JSON code for [Globalpay](https://github.com/opencart/opencart/blob/master/upload/catalog/view/theme/default/template/extension/payment/globalpay_remote.tpl)/Sagepay/Realex payment gateways currently results in a locked(disabled) checkout button where authentication fails.  This seems to be caused by JSON ['ACSURL'](http://take.ms/hpH8K) returning a [null value](http://take.ms/qucp4) where the user fails to complete the payment fields, correctly.

Similar issues have been reported previously where failure to click Terms & Conditions (e.g. issue #3885 ) but this issue with JSON 'ACSURL' returning Null values looks unique to the Globalpay/Realex/Sagepay payment modules.

**Comment by chris-wm** [comment]
@TerryMGray Hi, I can replicate this issue if I manually set the $json variable in the SagePay controller file to null but I cannot replicate using the normal flow of the system. Your second link seems to imply this is happening on the terms and conditions step, is this correct? Are you using a one page checkout extension of any type, or anything that modifies the checkout process?
Regards

**Comment by TerryMGray** [comment]
@chris-wm yes, should have mentioned I am using a one-page checkout. It is Journal theme but the controller does use the default Realex/Globalpay controller at catalogue/controller/payment/realex-remote.php (/globalpay-remote.php). I should also mention the payment method used is Realex remote as I see the Sagepay module handles retrieval of $json['ACSURL'] differently.  

I'm not entirely sure where $verify_3ds->url is returned from but, from testing, I can populate it with Null by sending incomplete data to the payment provider (e.g. missing cardholder). Public function send() hardcodes $verify_3ds->url depending on the environment (Test, Live, etc.) but neither Realex Remote nor Globalpay Remote provide a way to populate the $json variable (as far as I can tell).

**Comment by chris-wm** [comment]
@TerryMGray Hi, there is a good chance that this is caused by the one-page checkout, I was struggling to replicate on a default store. Can you either speak to the extension developer to see if they are aware of similar problems or please raise a ticket on our [support site](https://help.welfordmedia.co.uk/),
regards

**Comment by TerryMGray** [comment]
@chris-wm, yes on testing with default opencart installation this does appear to relate specifically to the Quick Checkout. I have referred the matter to them for investigation. Thanks.

**Comment by TerryMGray** [comment]
@chris-wm, on further investigation $json['ACSURL'] does only return Null when the theme is used. The 'Confirm' checkout button freezes when Realex payments is used with the default opencart installation, but does not return Null.   I have referred the issue to the theme developer but it may be worthwhile checking Realex Remote payments with default opencart store for invalid card values. 

![checkout-screenshot](https://cloud.githubusercontent.com/assets/1483340/17627645/e4f0b53c-60a9-11e6-8f96-1b8afd418577.jpg)

I have raised a ticket on your support site but did note that Realex is not mentioned specifically, so please disregard if Realex payment module is outside your scope.

**Comment by TerryMGray** [comment]
@chris-wm, on further testing, I believe the issue with Realex Remote lies with public function send() in /catalogue/controller/payment.php which is not validating card values (name, card number, etc.) fully prior to posting $enc_data. That is $enc_data can send data which does not conform to Visa's 3d secure schema.

**Comment by chris-wm** [comment]
@TerryMGray Hi, I have tested the Realex remote on the latest version of OpenCart and if you try to add incorrect card details then it returns details of the error dynamically on the page. The confirm button is then reset once this message is displayed, it sounds like something is interrupting this process on your site, presumably the one page checkout,
regards

**Comment by TerryMGray** [comment]
Hi @chris-wm. The issue I am experiencing is not with invalid values, but rather empty values, such as missing cardholder name or CVV number. I have experienced the issue with default opencart settings - see screenshot above, although would add that I am running opencart v.2.0.3.1 rather than the latest release.  I have implemented a fix, although this is specific to the Quick Checkout used by the template  I am running. If you are unable to reproduce the issue when using Realex with empty payment card form values / 3d secure enabled (which can only be tested with a live Realex account) then I would suggest this issue should remain closed.
